wicket-core/src/main/java/org/apache/wicket/markup/head/filter/CspNonceHeaderResponse.java

```diff
@@ -0,0 +1,348 @@
+            CssUrlReferenceHeaderItem headerItem = (CssUrlReferenceHeaderItem) item;
+            item = new CssUrlReferenceWithNonceHeaderItem(
+                    headerItem.getUrl(),
+                    headerItem.getMedia(),
+                    headerItem.getRel()
+            ).setNonce(nonce);
+        }
+
+        super.render(item);
+    }
+
+    protected abstract String getNonce();
+
+    protected static void renderScriptReferenceHeaderItem(
+            Response response,
+            final CharSequence url,
+            final String id,
+            boolean defer,
+            String charset,
+            boolean async,
+            String nonce
+    ) {
+        Args.notEmpty(url, "url");
+        Args.notEmpty(nonce, "nonce");
+        boolean isAjax = RequestCycle.get().find(IPartialPageRequestHandler.class).isPresent();
+        // the url needs to be escaped when Ajax, because it will break the Ajax Response XML (WICKET-4777)
+        CharSequence escapedUrl = isAjax ? Strings.escapeMarkup(url): url;
+
+        HeaderItemAttributeMap attributes = new HeaderItemAttributeMap();
+        attributes.add(HeaderItemAttribute.SCRIPT_SRC, String.valueOf(escapedUrl));
+        attributes.add(HeaderItemAttribute.TYPE, "text/javascript");
+        if (id != null) {
+            attributes.add(HeaderItemAttribute.ID, id);
+        }
+        if (defer) {
+            attributes.add(HeaderItemAttribute.SCRIPT_DEFER, "defer");
+        }
+        if (async) {
+            attributes.add(HeaderItemAttribute.SCRIPT_ASYNC, "async");
+        }
+        attributes.add(HeaderItemAttribute.CSP_NONCE, nonce);
+        if (!Strings.isEmpty(charset)) {
+            attributes.add("charset", charset);
+        }
+
+        JavaScriptUtils.writeJavaScriptUrl(response, attributes);
+    }
+
+    protected static void renderCssReferenceHeaderItem(
+            Response response, String id, String url, String media, String rel, String nonce
+    ) {
+        Args.notEmpty(url, "url");
+
+        HeaderItemAttributeMap attributes = new HeaderItemAttributeMap();
+        if (Strings.isEmpty(rel) == false)
+        {
+            attributes.add(HeaderItemAttribute.LINK_REL, rel);
+        } else {
+            attributes.add(HeaderItemAttribute.LINK_REL, "stylesheet");
+        }
+
+        attributes.add(HeaderItemAttribute.LINK_HREF, String.valueOf(Strings.escapeMarkup(url)));
+
+        if (Strings.isEmpty(media) == false)
+        {
+            attributes.add(HeaderItemAttribute.LINK_MEDIA, media.toString());
+        }
+        if (Strings.isEmpty(id) == false)
+        {
+            attributes.add(HeaderItemAttribute.ID, id);
+        }
+
+        attributes.add(HeaderItemAttribute.CSP_NONCE, nonce);
+
+        CssUtils.writeLinkUrl(response, attributes);
+
+        response.write("\n");
+    }
+
+    private static class JavaScriptContentWithNonceHeaderItem extends JavaScriptContentHeaderItem {

```

Comment:
Just pull-up #nonce to JavaScriptContentHeaderItem, it's an official attribute so no reason not to have it there.